RockLib.Messaging.SQS/SQSReceiver.cs

```diff
@@ -327,8 +327,7 @@ string GetErrorMessage()
                     continue;
                 }
 
-                var tasks = response.Messages.Select(HandleAsync);
-                await Task.WhenAll(tasks).ConfigureAwait(false);
+                response.Messages.ForEach(x => _ = HandleAsync(x));
```

Comment:
I have concerns with this approach. I realize this is a draft PR, so maybe this is just trying to show the desired end state.

* We definitely do not want to replace our current approach with this, as it has the potential of breaking existing clients. Having some kind of option that lets a developer specifiy if messages should be processed concurrently or sequentially. The option should default to the current behavior.
* If we do this, we should use `ForEachAsync(...)` and await the `HandleAsync()` call. Note that `ForEachAsync()` is only in .NET 6 and above, so we would need to do this manually for .NET 4.8.

Reply:
> I have concerns with this approach. I realize this is a draft PR, so maybe this is just trying to show the desired end state.

Yea, my intention was just to show possible changes.

> * We definitely do not want to replace our current approach with this, as it has the potential of breaking existing clients. Having some kind of option that lets a developer specifiy if messages should be processed concurrently or sequentially. The option should default to the current behavior.

I agree, although I'd be surprised if anyone using this is under the impression that the messages are being processed consecutively.

> * If we do this, we should use `ForEachAsync(...)` and await the `HandleAsync()` call. Note that `ForEachAsync()` is only in .NET 6 and above, so we would need to do this manually for .NET 4.8.

Are you referring to `Parallel.ForEachAsync`? That might not be desired because, IIRC, the `Parallel` API was designed for CPU-bound processing (saturates the CPU with its own threads) whereas the typical call to an `async` method (as done in `SQSReceiver`) is designed for IO-bound processing (will use the TPL's thread-pool, a la green threads).

Our case is definitely IO-bound processing. I'm also under the impression that event-handling patterns are generally IO-bound, with CPU-bound work being explicitly put through some other API as needed (maybe Parallel).

Reply:
I think we agree that the default should not break, but that an option should be surfaced to override the behavior.

As for awaiting the call to `HandleAsync`, I'm drawing a blank on how to do that without serializing the message processing (AKA the way it is now).

Reply:
My understanding of `HandleAsync` is that it already catches and delegates exceptions, so there's no need to `await` it in terms of unwrapping and handling exceptions. It should be "safe" to have the TPL thread pool run the task when it can so `ReceiveMessages` can continue processing more messages from the queue.